upb/mini_table.c

```diff
@@ -0,0 +1,937 @@
+                                          upb_ToBase92(63));
+  in->oneof_state = kUpb_OneofState_EmittedOneofField;
+  return ptr;
+}
+
+const upb_MiniTable_Field* upb_MiniTable_FindFieldByNumber(
+    const upb_MiniTable* table, uint32_t number) {
+  int n = table->field_count;
+  for (int i = 0; i < n; i++) {
+    if (table->fields[i].number == number) {
+      return &table->fields[i];
+    }
+  }
+  return NULL;
+}
+
+/** Data decoder **************************************************************/
+
+typedef enum {
+  kUpb_LayoutItemType_Field,       // Non-oneof field data.
+  kUpb_LayoutItemType_OneofField,  // Oneof field data.
+  kUpb_LayoutItemType_OneofCase,   // Oneof case.
+} upb_LayoutItemType;
+
+#define kUpb_LayoutItem_IndexSentinel ((uint16_t)-1)
+
+typedef struct {
+  // Index of the corresponding field.  When this is a oneof field, the field's
+  // offset will be the index of the next field in a linked list.
+  uint16_t field_index;
+  upb_FieldRep rep;
+  upb_LayoutItemType type;
+
+  // Used for temporary storage while assigning offsets (internal only).
+  uint16_t offset;
+} upb_LayoutItem;
+
+typedef struct {
+  upb_LayoutItem* data;
+  size_t size;
+  size_t capacity;
+} upb_LayoutItemVector;
+
+typedef struct {
+  const char* end;
+  upb_MiniTable* table;
+  upb_MiniTable_Field* fields;
+  upb_MiniTablePlatform platform;
+  upb_LayoutItemVector vec;
+  upb_Arena* arena;
+  upb_Status* status;
+  jmp_buf err;
+} upb_MtDecoder;
+
+UPB_PRINTF(2, 3)
+UPB_NORETURN static void upb_MtDecoder_ErrorFormat(upb_MtDecoder* d,
+                                                   const char* fmt, ...) {
+  va_list argp;
+  upb_Status_SetErrorMessage(d->status, "Error building mini table: ");
+  va_start(argp, fmt);
+  upb_Status_VAppendErrorFormat(d->status, fmt, argp);
+  va_end(argp);
+  UPB_LONGJMP(d->err, 1);
+}
+
+static void upb_MtDecoder_CheckOutOfMemory(upb_MtDecoder* d, const void* ptr) {
+  if (!ptr) upb_MtDecoder_ErrorFormat(d, "Out of memory");
+}
+
+// In each field's offset, we temporarily store a presence classifier:
+enum PresenceClass {
+  kNoPresence = 0,
+  kHasbitPresence = 1,
+  kRequiredPresence = 2,
+  kOneofBase = 3,
+  // Negative values refer to a specific oneof with that number.  Positive
+  // values >= kOneofBase indicate that this field is in a oneof, and specify
+  // the next field in this oneof's linked list.
+};
+
+static const char* upb_MiniTable_DecodeBase92Varint(upb_MtDecoder* d,
+                                                    const char* ptr,
+                                                    char first_ch, uint8_t min,
+                                                    uint8_t max,
+                                                    uint32_t* out_val) {
+  uint32_t val = 0;
+  uint32_t shift = 0;
+  char ch = first_ch;
+  while (1) {
+    uint32_t bits = upb_FromBase92(ch) - upb_FromBase92(min);
+    val |= bits << shift;
+    if (ptr == d->end || *ptr < min || max < *ptr) {
```

Comment:
you remembered my penchant for directionality in comparisons <3

Reply:
I try. :)